### ISSUE TITLE
`opaque_over_existential` breaks Codable and other conformances

### ISSUE THREAD

**Issue by ZevEisenberg** [realm/SwiftLint#6046]
### New Issue Checklist

- [x] I've Updated SwiftLint to the latest version.
- [x] I've searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues).

### Bug Description

There are some cases where it is invalid to convert `any` to `some`, because it breaks protocol conformances. A minimum triggering example would be:

```diff
- init(from decoder: any Decoder) throws
+ init(from decoder: some Decoder) throws
```

To see why this is a problem, consider this file, called CodableLintBug.swift:

```swift
import Foundation

struct MyStruct {
  let name: String
}

extension MyStruct: Codable {
  enum CodingKeys: String, CodingKey {
    case name
  }
  // this "any" should not become "some"
  init(from decoder: any Decoder) throws {
    print(Self.self, #function, "customized implementation")
    let container = try decoder.container(keyedBy: CodingKeys.self)
    self.name = try container.decode(String.self, forKey: .name)
  }

  // this "any" should not become "some"
  func encode(to encoder: any Encoder) throws {
    print(Self.self, #function, "customized implementation")
    var container = encoder.container(keyedBy: CodingKeys.self)
    try container.encode(name, forKey: .name)
  }
}

let myStruct = MyStruct(name: "asdf")
let data = try JSONEncoder().encode(myStruct)
let decoded = try JSONDecoder().decode(MyStruct.self, from: data)
print("decoded \(decoded)")
```

When you run it with `swift CodableLintBug.swift`, you get this output:

```
MyStruct encode(to:) customized implementation
MyStruct init(from:) customized implementation
decoded MyStruct(name: "asdf")
```

However, if you enable the `opaque_over_existential` SwiftLint rule, it converts `any Decoder` and `any Encoder` to `some Decoder` and `some Encoder`, respectively. Running it again, it produces:

```
decoded MyStruct(name: "asdf")
```

Notice that it is still compiling without warnings or errors, but it is skipping over the customized conformances and instead going with the compiler-generated conformances. This could lead to some very confusing bugs! (This is also why people have proposed adding a keyword that indicates your intent to satisfy a protocol requirement, but that's for another discussion.)

⚠  this does not _only_ affect Codable/Decodable. It affects any function that takes an `any` parameter as part of satisfying a protocol conformance. I'm not sure if there's a way to detect this at a syntactic level, so I don't know if there's a completely correct way to implement this rule 🤔

```bash
$ swiftlint --no-cache --autocorrect
Correcting Swift files in current working directory
Correcting 'CodableLintBug.swift' (1/1)
/Users/zev/Downloads/CodableLintBug/CodableLintBug.swift: Corrected opaque_over_existential 2 times
Done correcting 1 file!
```

### Environment

* SwiftLint version: **0.59.0**
* Xcode version: **Xcode 16.3 (16E140) on macOS 15.4 (24E248)**
* Installation method used: **Homebrew and also install from standalone binaries provided on release page.**
* Configuration file:

```yml
opt_in_rules:
  - opaque_over_existential
```

Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? **No.**

**Comment by SimplyDanny** [comment]
Producing code that doesn't compile is one thing. Causing wrong code that compiles is nasty, however.

I noticed the issue with protocol conformances [in the PR](https://github.com/realm/SwiftLint/pull/5915#issuecomment-2564802032) that introduced the rule. In my example, though, the code wouldn't compile after the "fix".

The rule could ignore `any Encoder` and `any Decoder`, but that's not sufficient as you warned.

> This is also why people have proposed adding a keyword that indicates your intent to satisfy a protocol requirement, but that's for another discussion.

I've also been advocating for this keyword. It would be so helpful in many ways, even if it's added as optional only.

Not sure what to do with the rule now. It's impossible to decide if `some` is okay on the level of syntax only. Normally, an **opt-in** rule causing false positives is acceptable. The matter of wrong-compilable code concerns me, though. 🤔

**Comment by SimplyDanny** [comment]
There are a few options (or any combination of them):

1. Be very clear about the shortcomings of this rule in its rational.
1. Disable automatic fixes.
1. Exclude a few known types and function.
1. Remove the rule entirely (and maybe reimplement it later as an Analyzer rule).

What do you guys - @ZevEisenberg, @twoertwein, @sindresorhus, @weakfish - think?

**Comment by weakfl** [comment]
4.

**Comment by ZevEisenberg** [comment]
[Partly answering my own question](https://github.com/realm/SwiftLint?tab=readme-ov-file#analyze), but I don't think this answers the second part about protocol witnesses or conformances or whatever the right wording is. But I'm guessing the answer is no, unless it gives us access to all the protocols that a type professes conformance of? And then we'd have to see if the function we want to "fix" is part of a protocol (and its parent protocols, etc), and would we have access to _that_ information???

**Comment by SimplyDanny** [comment]
> [Partly answering my own question](https://github.com/realm/SwiftLint?tab=readme-ov-file#analyze), but I don't think this answers the second part about protocol witnesses or conformances or whatever the right wording is. But I'm guessing the answer is no, unless it gives us access to all the protocols that a type professes conformance of? And then we'd have to see if the function we want to "fix" is part of a protocol (and its parent protocols, etc), and would we have access to _that_ information???

This is my "maybe" in option 4 as I'm also not sure about that.

**Comment by ZevEisenberg** [comment]
I have a hard time voting for anything other than option 4. You'd have to understand so much nuance to use this rule effectively that, at that point, you would be equipped to just search the code yourself, occasionally, for `any [A-Z]` and seeing if anything new pops up.

**Comment by SimplyDanny** [comment]
> This may go into the void since the issue is closed but wanted to add a vote for later implementation as an analyzer rule.

This is definitely not going into void, @Jason-Abbott! 😉 
 
> I found `opaque_over_existential` quite instructive. I landed after here after searching for more information about the rule, to see if I could ignore `any Encoder/Decoder`, so I can certainly see how it's problematic with auto-fixes. But I would be glad for a non-auto-fixing hint like this.

This is a quite interesting perspective. There is currently only `warning` and `error`. Something like `info`, `note` or `remark` isn't either supported by SwiftLint nor Swift, nor Xcode to my knowledge.

Reporting a warning at least might appear important enough for people to "fix" it, even if the message indicates only a note. So there is always the danger that things break on the rule's suggestion.

As Zev suggested, searching for `any [A-Z]` intentionally every now and then might be the safer alternative until we find a way for either an Analyzer rule or a clearer reporting style.
